routers/api/v1/repo/compare.go

```diff
@@ -0,0 +1,443 @@
+	headGitRepo *git.Repository,
+	compareInfo *git.CompareInfo,
+	baseBranch, headBranch string,
+	baseIs, headIs map[string]bool) (string, string, *gitdiff.Diff, int, string, bool, bool) {
+
+	var (
+		repo = ctx.Repo.Repository
+		err  error
+
+		// return
+		headCommitID string
+		baseCommitID string
+		diff         *gitdiff.Diff
+		commitCount  int
+		headTarget   string
+	)
+
+	// Get diff information.
+	headCommitID = headBranch
+	if !headIs["commit"] {
+		if headIs["tag"] {
+			headCommitID, err = headGitRepo.GetTagCommitID(headBranch)
+		} else {
+			headCommitID, err = headGitRepo.GetBranchCommitID(headBranch)
+		}
+		if err != nil {
+			ctx.Error(http.StatusInternalServerError, "GetRefCommitID", err)
+			return "", "", nil, 0, "", false, true
+		}
+	}
+
+	headTarget = path.Join(headUser.Name, repo.Name)
+
+	if headCommitID == compareInfo.MergeBase {
+		return headCommitID, compareInfo.MergeBase, nil, 0, headTarget, true, false
+	}
+
+	diff, err = gitdiff.GetDiffRange(models.RepoPath(headUser.Name, headRepo.Name),
+		compareInfo.MergeBase, headCommitID, setting.Git.MaxGitDiffLines,
+		setting.Git.MaxGitDiffLineCharacters, setting.Git.MaxGitDiffFiles)
+	if err != nil {
+		ctx.Error(http.StatusInternalServerError, "GetDiffRange", err)
+		return "", "", nil, 0, "", false, true
+	}
+
+	baseGitRepo := ctx.Repo.GitRepo
+	baseCommitID = baseBranch
+	if !baseIs["commit"] {
+		if baseIs["tag"] {
+			baseCommitID, err = baseGitRepo.GetTagCommitID(baseBranch)
+		} else {
+			baseCommitID, err = baseGitRepo.GetBranchCommitID(baseBranch)
+		}
+		if err != nil {
+			ctx.Error(http.StatusInternalServerError, "GetRefCommitID", err)
+			return "", "", nil, 0, "", false, true
+		}
+	}
+
+	commitCount = compareInfo.Commits.Len()
+
+	return headCommitID, baseCommitID, diff, commitCount, headTarget, false, false
+}
+
+func toCommits(ctx *context.APIContext, compare *api.Compare, compareInfo *git.CompareInfo, commitCount int) error {
+	compare.Commits = make([]*api.Commit, 0, commitCount)
+	compare.CommitCount = commitCount
+
+	// From list to array.
+	for cur := compareInfo.Commits.Front(); cur != nil; cur = cur.Next() {
+		// Convert git.Commit to api.Commit.
+		commit, err := toCommit(ctx, ctx.Repo.Repository, cur.Value.(*git.Commit), nil)
+		if err != nil {
+			return err
+		}
+		compare.Commits = append(compare.Commits, commit)
+	}
+	return nil
+}
+
+func toDiff(ctx *context.APIContext, compare *api.Compare, diff *gitdiff.Diff) error {
+	compare.Diff = new(api.Diff)
+	compare.Diff.TotalAddition = diff.TotalAddition
+	compare.Diff.TotalDeletion = diff.TotalDeletion
+	for i := range diff.Files {
+		compareDiffFile, err := toCompareDiff(ctx, diff.Files[i])
+		if err != nil {
+			return err
+		}
+		compare.Diff.Files = append(compare.Diff.Files, compareDiffFile)
+	}
+	compare.Diff.IsIncomplete = diff.IsIncomplete
+	return nil
+}
+
+func toCompareDiff(ctx *context.APIContext, compareDiffFile *gitdiff.DiffFile) (*api.DiffFile, error) {
```

Comment:
type convertions expecialy for API should go into `modules/convert/....go`

Reply:
This method of organization is modeled after the file routers/api/v1/repo/commits.go, and it is not just a conversion structure, but also other logic.

Reply:
convertion logic also should went to convert ... make it reusable :)